src/AutoMapper/Internal/ExpressionFactory.cs

```diff
@@ -56,18 +56,26 @@ public static Expression ForEach(Expression collection, ParameterExpression loop
 
             var breakLabel = Label("LoopBreak");
 
+            var disposableVar = Variable(typeof(IDisposable), "disposable");
+            Expression tryDisposeCall = Block(new[] { disposableVar },
+                Assign(disposableVar, TypeAs(enumeratorVar, typeof(IDisposable))),
```

Comment:
This is a boxing conversion. Also an information known beforehand because we can check `enumeratorType` and check for a `Dispose` method. If it exists, add the call.

Reply:
How can you check the `enumeratorType` for a `Dispose` method?

I tried it, but if the `GetEnumerator` method returns `IEnumerator` (instead of `IEnumerator<T>`) but the runtime enumerator is of type `IEnumerator<T>` and hance `IDisposable` you better check the concrete type at runtime.

I do not know of any way to do the runtime check without boxing.

Reply:
Don't check the runtime type. Just use `enumeratorType`. In a proper implementation the return type of `GetEnumerator` should be a more specific type.

Reply:
```
public class Demo : IEnumerable
{
    public IEnumerator GetEnumerator()
    {
        return new EnumeratorDisposable();
        return new EnumeratorOfT();
    }

    private class EnumeratorDisposable : IEnumerator, IDisposable
    {
        public bool MoveNext() => true;

        public void Reset() { }

        public object Current { get; }

        public void Dispose() { }
    }

    private class EnumeratorOfT : IEnumerator<object>, IDisposable
    {
        public bool MoveNext() => true;

        public void Reset() { }

        public object Current { get; }

        public void Dispose() { }
    }
}

```
In these two cases, the `Dispose` method would not be called.
That is not what I want.

Reply:
Having a struct as enumerator is a well known optimization, used throughout the BCL. Your code effectively throws that away. Check the IL generated by the compiler and you'll see there's no boxing there.

Reply:
Oh, i did not think of `struct`s.

But, as far as I can see, the IL generated only optimizes the boxing away in case there is an `IEnumerator<T>`. In case if an `IEnumerator`, it casts and boxes (like me).

Maybe I can make two different loops:

- a loop optimized for `IEnumerator<T>`
- a loop with runtime check for `IEnumerator`

What do you think?

Reply:
I updated the PR. Can you please check it again. Thank you.

The missing tests will be added as soon as the changes in the ForEach method are ok for you.